op-service/eth/blob.go

```diff
@@ -70,60 +72,192 @@ func VerifyBlobProof(blob *Blob, commitment kzg4844.Commitment, proof kzg4844.Pr
 
 // FromData encodes the given input data into this blob. The encoding scheme is as follows:
 //
-// First, field elements are encoded as big-endian uint256 in BLS modulus range. To avoid modulus
-// overflow, we can't use the full 32 bytes, so we write data only to the topmost 31 bytes of each.
-// TODO: we can optimize this to get a bit more data from the blobs by using the top byte
-// partially.
+// In each round we perform 7 reads of input of lengths (31,1,31,1,31,1,31) bytes respectively for
+// a total of 127 bytes. This data is encoded into the next 4 field elements of the output by
+// placing each of the 4x31 byte chunks into bytes [1:32] of its respective field element. The
+// three single byte chunks (24 bits) are split into 4x6-bit chunks, each of which is written into
+// the top most byte of its respective field element, leaving the top 2 bits of each field element
+// empty to avoid modulus overflow.  This process is repeated for up to 1024 rounds until all data
+// is encoded.
 //
-// The first field element encodes the length of input data as a little endian uint32 in its
-// topmost 4 (out of 31) bytes, and the first 27 bytes of the input data in its remaining 27
-// bytes.
-//
-// The remaining field elements each encode 31 bytes of the remaining input data, up until the end
-// of the input.
-//
-// TODO: version the encoding format to allow for future encoding changes
+// For only the very first output field, bytes [1:5] are used to encode the version and the length
+// of the data.
 func (b *Blob) FromData(data Data) error {
 	if len(data) > MaxBlobDataSize {
```

Comment:
There are 2 checks for the size, are both necessary?

Reply:
Both error messages are the same so its not possible to distinguish a difference between the 2 cases. If both are required, we should modify the error messages so they are distinct and then test for both cases

Reply:
the second isn't necessary, was more for documentation that the high order byte will never be set and we can always squeeze this into 3 bytes. removed & replaced with a comment.